firedrake/slate/slac/compiler.py

```diff
@@ -65,159 +195,253 @@ def compile_expression(slate_expr, tsfc_parameters=None):
-    # coefficient
+    if builder.needs_mesh_layers:
+        # In the presence of interior horizontal facet calls, an
+        # IF-ELIF-ELSE block is generated using the mesh levels
+        # as conditions for which calls are needed:
+        #
+        #    IF (layer == bottom_layer):
+        #        *bottom calls
+        #    ELSE IF (layer == top_layer):
+        #        *top calls
+        #    ELSE:
+        #        *top calls
+        #        *bottom calls
+        #
+        # Any extruded top or bottom calls for extruded facets are
+        # included within the appropriate mesh-level IF-blocks. If
+        # no interior horizontal facet calls are present, then
+        # standard IF-blocks are generated for exterior top/bottom
+        # facet calls when appropriate:
+        #
+        #    IF (layer == bottom_layer):
+        #        *bottom calls
+        #
+        #    IF (layer == top_layer):
+        #        *top calls
+        #
+        # The mesh level is an integer provided as a macro kernel
+        # argument.
+
+        # FIXME: No variable layers assumption
+        statements.append(ast.FlatBlock("/* Mesh levels: */\n"))
+        num_layers = slate_expr.ufl_domain().topological.layers - 1
+        int_top = builder.assembly_calls["interior_facet_horiz_top"]
+        int_btm = builder.assembly_calls["interior_facet_horiz_bottom"]
+        ext_top = builder.assembly_calls["exterior_facet_top"]
+        ext_btm = builder.assembly_calls["exterior_facet_bottom"]
+
+        if int_top + int_btm:
+            statements.append(ast.FlatBlock("/* Interior and top/bottom calls */\n"))
+            # NOTE: The "top" cell has a interior horizontal facet
+            # which is locally on the "bottom." And vice versa.
+            top_calls = int_btm + ext_top
+            btm_calls = int_top + ext_btm
+
+            block_else = ast.Block(int_top + int_btm, open_scope=True)
+            block_top = ast.Block(top_calls, open_scope=True)
+            block_btm = ast.Block(btm_calls, open_scope=True)
+
+            elif_block = ast.If(ast.Eq(mesh_layer_sym, num_layers - 1),
+                                (block_top, block_else))
+
+            statements.append(ast.If(ast.Eq(mesh_layer_sym, 0),
+                                     (block_btm, elif_block)))
+        else:
+            if ext_btm:
+                statements.append(ast.FlatBlock("/* Bottom calls */\n"))
+                layer = 0
+                block_btm = ast.Block(ext_btm, open_scope=True)
+                statements.append(ast.If(ast.Eq(mesh_layer_sym, layer),
+                                         (block_btm,)))
+
+            if ext_top:
+                statements.append(ast.FlatBlock("/* Top calls */\n"))
+                layer = num_layers - 1
+                block_top = ast.Block(ext_top, open_scope=True)
+                statements.append(ast.If(ast.Eq(mesh_layer_sym, layer),
+                                         (block_top,)))
```

Comment:
This kind of repetition of the same code looks to ripe for simplification.  What about:

```python
bottom = ast.Block(int_top + ext_btm, open_scope=True)
top = ast.Block(int_btm + ext_top, open_scope=True)
rest = ast.Block(int_btm + int_top, open_scope=True)
statements.append(ast.If(ast.Eq(mesh_layer_sym, 0), 
                         bottom,
                         ast.If(ast.Eq(mesh_layer_sym, num_layers - 1),
                                top,
                                rest)))
```

Reply:
Having an empty statement in the generated code might look a little less pretty, but the compiler will elide it, so it'll be fine.